src/vanilla.ts

```diff
@@ -107,23 +107,29 @@ const buildProxyFunction = (
       : Object.create(Object.getPrototypeOf(target))
     markToTrack(snap, true) // mark to track
     snapCache.set(target, [version, snap])
-    Reflect.ownKeys(target).forEach((key) => {
-      const value = Reflect.get(target, key)
-      if (refSet.has(value as object)) {
-        markToTrack(value as object, false) // mark not to track
-        snap[key] = value
-      } else if (value instanceof Promise) {
-        Object.defineProperty(snap, key, {
-          get() {
-            return handlePromise(value)
-          },
-        })
-      } else if (proxyStateMap.has(value as object)) {
-        snap[key] = snapshot(value as object, handlePromise)
-      } else {
-        snap[key] = value
+    const enumerable = true
+    ;[...Reflect.ownKeys(target), ...findPrototypeGetters(target)].forEach(
```

Comment:
Fwiw @dai-shi I pushed up this change which always uses `Object.defineProperty`, so the diff is much smaller.

Not sure if that will help your concert wrt the complexity increase in the code, but it was easy to try.